ipc4: implement mtrace logging support

kv2019i opened the pull request:
Set of patches to implement mtrace support in SOF firmware.

Requires Zephyr and Zephyr PR https://github.com/zephyrproject-rtos/zephyr/pull/49147 to work.

V8 uploaded - ready for review now

kv2019i commented on src/audio/base_fw.c:
```diff
@@ -71,7 +81,11 @@ static int basefw_config(uint32_t *data_offset, char *data)
 	tuple = next_tuple(tuple);
-	set_tuple_uint32(tuple, IPC4_TRACE_LOG_BYTES_FW_CFG, 0);
+#ifdef CONFIG_LOG_BACKEND_ADSP_MTRACE
+	log_bytes_size = 1024;
```
Need to put this magic number somewhere (or actually, get from some runtime interface).

ujfalusi replied:
The thing with the LOG_BYTES definition: in some platforms it is set to the _whole_ size of the log window (1024 * number of slots), on some platforms it is set to the size of _one_ slot (1024).
The only consistency is that it is _set_ if mtrace type of log extraction is available, thus the kernel ignores the value itself and checks the slot mapping page to find the slots.

ujfalusi replied:
iow, setting it to 1024 is the right thing if mtrace is usable.

RanderWang replied:
#define MAILBOX_TRACE_SIZE	SRAM_TRACE_SIZE ?

kv2019i replied:
Ack, I'll pick some define for this. In theory this value should come from Zephyr side and code on SOF side should be hw agnostic.

kv2019i commented on src/audio/base_fw.c:
```diff
@@ -303,6 +317,38 @@ static int basefw_get_large_config(struct comp_dev *dev,
+	return IPC4_UNKNOWN_MESSAGE_TYPE;
+#endif
+}
+
```
Not entirely sure of the location. This is an interface of the basefw, but otherwise this code might be better in a separate file somewhere.

kv2019i replied:
I now moved this to a separate file as the amount of code kept increasing and it feels out of place in base_fw.c.

kv2019i commented on src/audio/base_fw.c:
```diff
@@ -314,6 +360,11 @@ static int basefw_set_large_config(struct comp_dev *dev,
 		tr_warn(&basefw_comp_tr, "returning success for Set FW_CONFIG without handling it");
 		return 0;
+	case IPC4_SYSTEM_TIME:
+		tr_warn(&basefw_comp_tr, "returning success for Set SYSTEM_TIME without handling it");
```
@ujfalusi Kernel will fail if we don't handle this. I don't really see a reason for us to set the system time, we'd have to reset the Zephyr kernel time here and I don't see a good reason to do that.

ujfalusi replied:
I would lover it to tr_dbg() level. The reason for this is to have the timestamps in sync (or close to in sync) with the kernel time.
When you look at both logs with times, you might be able to see better and match the events easier if the times are in sync.
But ignoring it is also fine, it is not an error, it is how it is implemented in SOF.

kv2019i replied:
Will do this.

lyakh replied:
presumably we could store the delta and then add it in logs?

kv2019i replied:
@lyakh It's just that the timestamps are added in Zephyr logging core, so we'd need an interface to feed this timing delta from app to the Zephyr logging subsystem. Might be doable, but I'd prefer to handle this is a follow-up task if needed.

kv2019i commented on src/ipc/ipc4/handler.c:
```diff
@@ -52,6 +52,10 @@ struct ipc4_msg_data msg_data;
 
+static bool ipc_cmd_in_progress;
+
+static struct ipc_msg msg_notify;
```
In theory we could use "msg_reply", but I was 100% sure this is safe, so added a separate object for the notifies. @ujfalusi @RanderWang @mmaka1 ?

ujfalusi replied:
In theory you could send the notification in between a message reception and sending back the reply, but why do you need this a global?

RanderWang replied:
msg_reply is for host msg, notification is for dsp msg, so we need separate msg. we don't need global msg for notification. There are a few types of notification and we can create msg for each type.

kv2019i commented on src/ipc/ipc4/handler.c:
```diff
@@ -975,6 +979,26 @@ void ipc_boot_complete_msg(struct ipc_cmd_hdr *header, uint32_t data)
+
+	/* TODO: need to find a cleaner solution */
+	if (msg_data.delayed_reply || ipc_cmd_in_progress)
+		return;
```
Ok this is the ugliest part. If a notification is sent while kernel is waiting for a response, that response is not correctly handled. Not sure if this is a kernel bug, or something we have to avoid as per the protocol. @ujfalusi @RanderWang @mmaka1 @marcinszkudlinski ?

ujfalusi replied:
Hrm, the kernel should be able to handle it and afaik I have seen it with closed source firmware. The tx+reply handling is detached from notification handling on the kernel side at least. It must be working.

RanderWang replied:
? is there any issue for bi-direction ipc msg ? I don't get it.   There are two sets of ipc register for tx and rx

kv2019i replied:
@ujfalusi @RanderWang Without this adhoc protection, I see BIND fail when a notification is send between BIND REQ received and before a response is sent back. I've not yet debugged why this happens and whether this is a problem on kernel or firmware side.

ujfalusi replied:
@kv2019i, do you have logs from kernel (with sof_debug=513)?

kv2019i replied:
@RanderWang @ujfalusi Mystery solved. The problem was caused by sending a notify from a different DSP core, not related to the ongoing request-response. Timing just happened to match with that. Will send an updated PR today.

kv2019i commented on tools/mtrace/mtrace-reader.py:
```diff
@@ -0,0 +1,26 @@
+# Tool to stream data from Linux SOF driver "mtrace" debugfs
+# interface to standard output. Plain "cat" is not sufficient
+# as each read() syscall returns log data with a 32bit binary
+# header, containing the payload length.
```
If we keep the current mtrace debugfs, we'll need a simple tool like this to stream data out. 

Note the log decoder can be run on the stdout of this tool.

ujfalusi replied:
For the record, the current debugfs implementation is:
- user space gives a buffer of size X
- kernel reads new data chunk from the slot up to (X - 4) bytes ( or capped at 1020 bytes due to slot size wrapping)
- kernel writes the length of the chunk as u32 to the start of the user buffer
- kernel writes the log data from offset 4 to user buffer
- kernel tell user space that it has exhausted the given buffer (tells that X bytes have been written)
- user space reads first u32 to know the log chunk size in buffer
- user space gives a new buffer since it was told that the previous has been filled up.
- and so on.

This is not `cat to stdout` friendly for sure, but easier to tool it and simpler to implement in kernel.

kv2019i replied:
Ack, let's keep this interface.

kv2019i reviewed (commented):
My comments inline on parts that still need attention before this can be undrafted.

kv2019i commented on src/audio/base_fw.c:
```diff
@@ -303,6 +317,38 @@ static int basefw_get_large_config(struct comp_dev *dev,
+	    k_uptime_delta(&msg_notify_last_sent) >= IPC4_MTRACE_NOTIFY_MAX_DELAY_MS)
+		ipc_send_buffer_status_notify();
+
+	msg_notify_last_sent = k_uptime_get();
```
Oops, needs to be in above "if()" statement.

RanderWang commented on src/audio/base_fw.c:
```diff
@@ -314,6 +360,11 @@ static int basefw_set_large_config(struct comp_dev *dev,
+		tr_warn(&basefw_comp_tr, "returning success for Set SYSTEM_TIME without handling it");
+		return 0;
+	case IPC4_ENABLE_LOGS:
+		return mtrace_enable();
```
we need mtrace_disable ?

kv2019i replied:
Ack, will add.

lgirdwood commented:
For the record, I'm looking for something hard coded and simple to get data out as the first merge point. We can then improve with dynamic config/tuning in subsequent updates.

lyakh commented on tools/mtrace/mtrace-reader.py:
```diff
@@ -0,0 +1,26 @@
+MTRACE_FILE = "/sys/kernel/debug/sof/mtrace/core0"
+
+fd = os.open(MTRACE_FILE, os.O_RDONLY)
+while fd >= 0:
```
I have no idea about .py. How does this `os.open()`, `os.read()` compare to `with open(MTRACE_FILE) as file:`?

kv2019i replied:
@lyakh  This is a feature of the debugfs interface. Every read system call returns as much data as the application requests for. The first 32bit tells how much of that data is actual log data. Due to this protocol, all higher level file reading abstractions will not be able to handle the debugfs interface. Instead we use direct os.read() calls without any intermediate buffering.

lgirdwood commented on src/ipc/ipc4/mtrace.c:
```diff
@@ -0,0 +1,128 @@
+#include <sof/ipc/common.h>
+#include <ipc4/base_fw.h>
+
+#undef panic // SOF headers may redefine panic
```
I will try and fix this panic redefinition this week.

lgirdwood commented on src/ipc/ipc4/mtrace.c:
```diff
@@ -0,0 +1,128 @@
+
+		mtrace_aging_timer = log_state->aging_timer_period;
+
+		schedule_task_init_ll(&mtrace_work,
```
Any reason why this does not use Zephyr native scheduling API ? Is the intention to run this atm as a task in the LL domain so that we have guaranteed DSP time to get trace out (and since the data can fill the buffer quickly).

I suppose this can be addressed once the scheduling API is cleaned up.

kv2019i replied:
@lgirdwood I tried first with k_timer and then with k_work, but the problem I face is that the callback comes in the wrong context. I could set up a k_thread pinned to core 0, and run a k_work there, but then I'm basicly rewriting the boilerplate we have in zephyr_ll.c already. I don't particularly like this, but this is the only version I could get to reliably work for me.

andyross replied:
What's the requirement that it run on core 0?  Seems like for "idle work" kind of things, symmetric scheduling is generally a better fit anyway (i.e. it will never impact latency of the high priority audio work).

But that said: it would be perfectly legal to pin the system work queue thread to a specific core in application code, and if SOF wants its low priority stuff always on core 0, then that seems like it would work.

kv2019i replied:
@andyross This is actually just an SOF-side application design choice. All the IPC code is designed to run on single core, with no protection for multicore access. That single core just happens to be 0. And my empirical tests confirmed that basic tests start failing if I allow the IPC notifications to be sent from other cores. So this is more of a "If the mountain will not come to Mohammed, Mohammed will go to the mountain" case. Either I make the IPC stack multicore safe, or I accept the status quo and send my notifies from the IPC core. :)

But, but, I'll probably make another attempt at using a native method for this today. One big drawback with using the SOF ll-scheduler interface is that it'll get called every 1ms. This is very wasteful as the aging timer for logging is typically configured to 250ms and I can see it could be much slower than that.

kv2019i replied:
Ok, @lgirdwood @andyross now implemented with a native k_thread and k_sleep. Wasted a few hours when I had accidentally set the priority to that of the idle thread and my thread was not scheduled at all when other parts of SOF were not active, but once I figure out my mistake, this started to work. I used thread pinning to ensure the thread only gets scheduled on the same code as IPC stack is running on. Still not super pretty, but fits the current design and makes it explicitly why it is done so. The code could still be improved by having a k_timer and reprogramming that every time  a notification is sent based on buffer fill level, but not sure this is worth the extra code anymore.

kv2019i commented:
V4 update:
- implemented the worker thread using native interfaces -- a bit more code, but this is more efficient as we don't need to run anything extra in the 1ms ll cadence
- added overlay changes to make this the default for all IPC4 platforms
- fixed a few bugs w.r.t. draining logs at end when streaming is stopped
- still keeping as draft as the Zephyr side PR is still in review

lyakh commented on src/ipc/ipc4/mtrace.c:
```diff
@@ -0,0 +1,172 @@
+{
+	if (arch_proc_id() == MTRACE_IPC_CORE &&
+	    (space_left < NOTIFY_BUFFER_STATUS_THRESHOLD ||
+	     k_uptime_delta(&mtrace_notify_last_sent) >= mtrace_aging_timer)) {
```
does this mean, that if core 1 logs a lot and core 0 doesn't log at all, we'll potentially lose those logs? Shouldn't we `k_wakeup()` if running on a different core and reaching the threshold?

kv2019i replied:
Yeah, the multicore part will require more work. The mtrace protocol actually defines separate slots of each core, but this is not part of this PR (plus with current windows set up by the bootloader, we don't have enough space for this approach). So this PR's approach to multicore is  a simple one. The logs won't be lost as the mtrace_thread (pinned on core0) will flush the logs for all cores. It does mean that in some multicore cases user-space has to set a lower aging timer value (this can be set via debugfs already now). 

In longer term, the current 4K slot is too little to relay logs for all cores, so we'll need a bigger rework, but that will be done in later PRs.

lyakh replied:
> The logs won't be lost as the mtrace_thread (pinned on core0) will flush the logs for all cores.

But if one or several secondary cores log more than the buffer can hold between two thread runs, which are currently set to 1s? You can log a lot in 1s.

kv2019i replied:
@lyakh Kernel sets this by default to 250ms. It's still a lot of time, so I think the eventual solution is to use the per-core buffers.

lyakh replied:
why would per-core buffers help? With 4 cores you'd be better off putting all the memory into one buffer than splitting it into 4 parts? But either way, doesn't increasing the buffer size just make this problem less likely to bite us instead of fixing it? The solution seems to be to make sure logging on secondary cores wakes up the logging thread if needed - therefore `k_wakeup()`

kv2019i replied:
The max buffer per core is 4K (as per protocol). But it is allowed to have a separate buffer for each core, so that increases total log buffering to 4x4K on a 4core system.

lyakh commented on src/ipc/ipc4/mtrace.c:
```diff
@@ -0,0 +1,172 @@
+static void mtrace_thread_fn(void *p1, void *p2, void *p3)
+{
+	while (1) {
+		if (k_uptime_delta(&mtrace_notify_last_sent) >= mtrace_aging_timer &&
```
if using `k_wakeup()` we'd have to drop this timeout check

kv2019i replied:
I think we need always a timer. The implementation could be improved by reprogramming the timer whenever new data is added to the buffer (if data is added at a constant fast enough pace, mtrace_thread_fn doesn't need to run ever), but this does not seem to be warranted and the simpler code does the job.

andyross replied:
Note that this is recapitulating very similar logic in use in the Zephyr logging subsystem when CONFIG_LOG_DEFERRED=y.   You could almost certainly freeload on top of that instead (IIRC your "process" callback happens in the log thread on a timer).  @nordic-krch would be the expert.

Threads are very expensive if all you want is a regular low-priority function called.  Another option would be to use a system work queue callback that just reschedules itself.

kv2019i replied:
@andyross Thanks! I now took a tour of the log-deferred code and yeah, I'm doing a double here. Hmm. I did look at system work queue, but I couldn't figure out how to ensure the callback would be pinned on the specific core (and it wouldn't really make sense for the system wq thread).  

Hmm. So the k_thread is a now fairly costly workaround for inability to use the IPC stack from multiple cores.

On cache coherent systems, it'd be obvious to just change the IPC code, but in this case there's always a cost. Starts to seem like a per-cpu wq would be very practical on these systems. We can of course do this at app (SOF) level the SOF LL scheduler in practise is just the thing, it creates core-pinned threads and provides an API to get work done on these pinned threads with domain specific scheduling order.

Maybe I need to switch back to earlier version and use the SOF pinned threads.

andyross replied:
Oh sorry, I didn't realize this was part of the same core 0 issue. :)

But actually the same ideas hold here too: that logging thread could absolutely be pinned to core 0 if that's what the application wants.  You could probably do it with a quick hack out of SOF code, but a kconfig like "CONFIG_LOG_THREAD_CPUMASK" would be easy to do and easy to support.

lyakh replied:
> I think we need always a timer.

sorry, I wasn't proposing to drop the thread, on the contrary - I was proposing to wake it up to send the notification from its context on core 0 always instead of potentially notifying from any other core where logging is called. Then you'd only send notifications from the thread and the hook would just wake it up.

lyakh replied:
> Threads are very expensive

@andyross hmm, it looks pretty with a thread though :-) Well, if it's really too expensive, a work would be good too, if it can be pinned to core 0... Although do we need that? Would anything break if we allow that work to run on any core?

andyross replied:
Heh, "too expensive" is an app decision.  But this thread in particular is being granted a 4k stack, vs. the (I think) 24 bytes required for a workq item.  (I'm sure it doesn't need that much, though getting xtensa calls onto stacks smaller than 512 bytes is very difficult; and in any case when we go to an MMU environment in the future 4k will be the mapping granularity anyway).

If you do want to pin the system work queue though, you can find the thread in the system work queue object, which is a public symbol in kernel.h, something like:

```
/* Do this early, before any items get added */
k_thread_cpu_pin(&k_sys_work_q.thread, 0);
```

But @lgirdwood is right, this isn't a -1 and there's no point in holding things up for what amount to

kv2019i replied:
@andyross Pinning the sys_work_q is tempting, but seems a bit evasive. We might have good reasons to schedule sys_work_q stuff on any available core.

I now took yet another approach and implemented the worker logic with SOF's EDF scheduling API. This basicly creates a k_thread pinned on a CPU anyways, so why not piggy back on this.

lyakh commented on src/ipc/ipc4/mtrace.c:
```diff
@@ -0,0 +1,172 @@
+	    (space_left < NOTIFY_BUFFER_STATUS_THRESHOLD ||
+	     k_uptime_delta(&mtrace_notify_last_sent) >= mtrace_aging_timer)) {
+		ipc_send_buffer_status_notify();
+		mtrace_notify_last_sent = k_uptime_get();
```
looks like locking is needed. What if the logging thread is run and has already notified the host, but here it's interrupted by an interrupt, which logs too and increments `mtrace_bytes_pending` after that we return to the thread, and it sets `mtrace_bytes_pending = 0`

kv2019i replied:
Ack, let me look at this. The mtrace_bytes_pending value is actually not sent to the host. The IPC just tells host that "there's something in the buffer" and host will go and check the read/write pointers directly from SRAM. So I think in this case it's enough we send one indication and host will also retrieve the logs written from interrupt. And I'm a bit worried about using spinlocks here as the log callbacks may come from e.g. panic handler.

lyakh replied:
right, but `mtrace_bytes_pending` is checked below in the thread, so if it's 0, the host will not be notified. Sure, you cannot spinlock around `ipc_send_buffer_status_notify()` but maybe something like this would work:
```
size_t bytes_reported = mtrace_bytes_pending;
ipc_send_buffer_status_notify();
spin_lock();
mtrace_bytes_pending -= bytes_reported;
spin_unlock();
```
and below you'd hold the same spinlock when adding to `mtrace_bytes_pending`

kv2019i replied:
@lyakh I checked and the logging backend is serializing the hook calls, so the above race cannot handle.

lyakh reviewed (changes requested):
nice! Just a couple of things to clean up

lgirdwood commented:
All, the important thing here is that we need some data out as the first step. Today we have no logging. We can improve our logging infra incrementally (and we can track this with FEATURES).

kv2019i commented:
V5 update:
 - improved the SOF EDF adaptation to Zephyr a bit and implemented the mtrace work thread logic using edf_schedule.h (this is basicly a k_thread pinned on a thread... and it's a cost we are already taking so why not piggyback on this)
 - bugfix to handle runtime suspend cases to mtrace-reader.py
 - still keeping as draft as the Zephyr part is not merged

kv2019i commented:
V6 update:
 - not included a manifest update to bring in the Zephyr dependencies
 - a few bugfixes to the EDF adaptation for recurring EDF tasks
 - added a more clear TODO for multicore; I spent a lot of time trying to use notifier.h to wake up a task from one core to another, but kept hitting with weird issues... this version now explicitly optimized for single-core usege. it will work wiht multicore, but one may need to use a more aggressive aging timer not to lose logs

lgirdwood commented:
@keqiaozhang @fredoh9 @marc-hb can we update the Zephyr SDK to align with the version used by Zephyr. Thanks

kv2019i commented:
V7 update:
- updated github workflows to use a specific zephyr docker-image that containts 0.15 SDK
- there is one IPC4 build failure left but that is a compiler warning handled via https://github.com/thesofproject/sof/pull/6211
- no change to code otherwise

kv2019i commented:
V8 update:
- fix XTOS IPC4 build caught by CI job "[checktree (IPC_VERSION=IPC4 UNSIGNED_list= SIGNED_list='tgl tgl-h cnl icl jsl')](https://github.com/thesofproject/sof/runs/8078564819?check_suite_focus=true#logs)"

abonislawski commented on src/include/ipc4/mtrace.h:
```diff
@@ -0,0 +1,18 @@
+
+#define SOF_MTRACE_PAGE_SIZE			0x1000
+
+int ipc4_set_enable_logs(bool first_block,
```
Not sure if this is a good place (mtrace files) for such function because IPC4_ENABLE_LOGS in BASE_FW is used to enable all types of log modules and for probe logs we need to have IPC4_ENABLE_LOGS & IPC4_SYSTEM_TIME too.

I didnt added anything for probes here (just return 0 for IPC4_ENABLE_LOGS) at the beginning but now I think we should have a generic function to enable logs

kv2019i replied:
True, and maybe I need to rename this to something more common.

kv2019i replied:
V9 uploaded, I moved this to ipc4/logging.h. I left a TODO in comments that we need to extend IPC4_ENABLE_LOGS to support configure the type of logging.

abonislawski commented on src/ipc/ipc4/mtrace.c:
```diff
@@ -0,0 +1,170 @@
+
+	dcache_invalidate_region((__sparse_force void __sparse_cache *)data, data_offset_or_size);
+
+	log_state = (struct ipc4_log_state_info *)data;
```
Actually this is not correct, for memory window log type this IPC will be extended and there is tracing_method information which we should use and then decide to run such mtrace init because in current form this will also trigger for probe log enable

kv2019i replied:
Ack @abonislawski . I enabled only the basic set as driver does not supported the extended struct yet.  FYI @ujfalusi we need to extend the driver side if we use the ENABLE_LOGS IPC to also enable the probes logging.

kv2019i replied:
@abonislawski ^^ I left a TODO to code for this, I don't have anything to test this against currently, so I prefer to start with something I can test.

abonislawski replied:
Thats absolutely fine, I can extend this later as I can easily test both types of this ipc message but please extend this in driver because extended message version is actually used for mtrace and normal (this one) version for probes :)

abonislawski reviewed (changes requested):
Very good addition to sof logs, the only thing missing is proper handling of IPC4_ENABLE_LOGS

kv2019i commented:
V9 update:
 - addressed @abonislawski 's revies and most notably reviewed mtrace.[ch] to logging.[ch] as the same IPC can be used for other logging backends
 - one bugfix to disabling logs that was causing IPC timeouts in testing
 - kept the SDK update... the "latest" tag still points to 0.14 SDK

lgirdwood commented:
Zephyr IPC4 build fix merged.
